airbyte-cdk/bulk/toolkits/load-s3/src/main/kotlin/io/airbyte/cdk/load/file/s3/S3MultipartUpload.kt

```diff
@@ -32,16 +38,34 @@ class S3MultipartUpload<T : OutputStream>(
             ?: throw IllegalStateException("Streaming upload part size is not configured")
     private val wrappingBuffer = streamProcessor.wrapper(underlyingBuffer)
 
-    inner class Writer : ObjectStorageStreamingUploadWriter {
-        override suspend fun write(bytes: ByteArray) {
-            wrappingBuffer.write(bytes)
-            if (underlyingBuffer.size() >= partSize) {
+    private val work = Channel<suspend () -> Unit>(Channel.UNLIMITED)
+
+    suspend fun start(): Job =
+        CoroutineScope(Dispatchers.IO).launch {
+            for (unit in work) {
                 uploadPart()
             }
+            completeInner()
+        }
+
+    inner class UploadWriter : Writer() {
+        override fun close() {
+            log.warn { "Close called on UploadWriter, ignoring." }
+        }
+
+        override fun flush() {
+            throw NotImplementedError("flush() is not supported on S3MultipartUpload.UploadWriter")
+        }
+
+        override fun write(str: String) {
+            wrappingBuffer.write(str.toByteArray(Charsets.UTF_8))
+            if (underlyingBuffer.size() >= partSize) {
+                runBlocking { work.send { uploadPart() } }
```

Comment:
You're right, we need to sequence the work items. The most straightforward thing to do is just to push whatever would have been in the body of the function into the channel.

```
    suspend fun run(block: suspend (OutputStream) -> Unit) = coroutineScope {
        log.info {
            "Starting multipart upload to ${response.bucket}/${response.key} (${response.uploadId}"
        }
        launch {
            for (item in workQueue) {
                item()
            }
            complete()
        }
        UploadStream().use {
            block(it)
        }
    }

    inner class UploadStream : OutputStream() {
        override fun close() = runBlocking {
            workQueue.send { workQueue.close() }
        }

        override fun flush() = runBlocking {
            workQueue.send { wrappingBuffer.flush() }
        }

        override fun write(b: Int) = runBlocking {
            workQueue.send {
                wrappingBuffer.write(b)
                if (underlyingBuffer.size() >= partSize) {
                    uploadPart()
                }
            }
        }

        override fun write(b: ByteArray) = runBlocking {
            workQueue.send {
                wrappingBuffer.write(b)
                if (underlyingBuffer.size() >= partSize) {
                    uploadPart()
                }
            }
        }
    }
```

This also simplifies the caller to

```
        val upload =
            S3MultipartUpload(
                client,
                response,
                ByteArrayOutputStream(),
                streamProcessor,
                uploadConfig
            )
        upload.run(block)
```

**Also note I converted the writer into an output stream. This will be necessary for Avro.**

Reply:
I should probably wrap those buffers in atomic references as well.